### PR TITLE
[Store] add structured object copy-mode policy

### PR DESCRIPTION
**Pull request by zxpdemonio**
## Motivation                                                                                                           
                                                                                                                          
  Some upper-layer frameworks may transfer very large tensors through the structured object interface. The zero-copy path 
  can fail when there is not enough registered buffer capacity, while these frameworks may not yet have proper handling for this failure mode. This change adds an explicit non-zero-copy transfer option as a fallback to avoid such failures. 
                                                                                                                          
  ## Description                                                                                                          
                                                                                                                          
  - Add `copy_mode` to `BundleTransferPolicy`:                                                                            
    - `auto`: preserve existing behavior, prefer zero-copy fast path and fallback when unavailable.                       
    - `zero_copy`: require the zero-copy `batch_put_from` path and fail if unsupported.                                   
    - `copy`: force the non-zero-copy `store.put` path.                                                                   
  - Expose `pre_registered_buffers` through `put_structured_object()`.                                                    
  - Validate pre-registered structured buffers to avoid treating copied, read-only, or non-contiguous buffers as already  
  registered.                                                                                                             
                                                                                                                          
  ## Test                                                                                                                 
                                                                                                                          
  ```bash                                                                                                                 
  PYTHONPATH=/root/Mooncake/mooncake-wheel \                                                                              
  /root/sglang-venv/bin/python -m pytest \                                                                                
  /root/Mooncake/mooncake-wheel/tests/test_structured_object_store.py                                                     
   ```                                                                                                                       
  Result:                                                                                                                 
                                                                                                                          
  33 passed                                                                                                               
                                                                                                                          
  Also checked:                                                                                                           
                                                                                                                          
  PYTHONPATH=/root/Mooncake/mooncake-wheel \                                                                              
  /root/sglang-venv/bin/python -m py_compile \                                                                            
  /root/Mooncake/mooncake-wheel/mooncake/structured_object_store.py \                                                     
  /root/Mooncake/mooncake-wheel/tests/test_structured_object_store.py


**Test results:**
- [x] Unit tests pass
- [x] Integration tests pass (if applicable)
- [x] Manual testing done (describe below)

## Module

- [ ] Transfer Engine (`mooncake-transfer-engine`)
- [x] Mooncake Store (`mooncake-store`)
- [ ] Mooncake EP (`mooncake-ep`)
- [ ] Mooncake PG (`mooncake-pg`)
- [ ] Integration (`mooncake-integration`)
- [ ] P2P Store (`mooncake-p2p-store`)
- [ ] Python Wheel (`mooncake-wheel`)
- [ ] Common (`mooncake-common`)
- [ ] Mooncake RL (`mooncake-rl`)
- [ ] CI/CD
- [ ] Docs
- [ ] Other

## Type of Change

- [ ] Bug fix
- [ ] New feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Performance improvement
- [x] Other


## Checklist

- [x] I have performed a self-review of my own code
- [x] I have formatted my code using `./scripts/code_format.sh`
- [x] I have run `pre-commit run --all-files` and all hooks pass
- [x] I have updated the documentation (if applicable)
- [x] I have added tests to prove my changes are effective
- [x] For changes >500 LOC: I have filed an RFC issue

## AI Assistance Disclosure



- [ ] No AI tools were used
- [x] AI tools were used (specify below)

Claude Code assisted with code exploration, implementation, formatting/build/test execution, and review of zero-sized tensor edge cases. I reviewed and validated the final changes.

### CODE REVIEW

File: mooncake-wheel/mooncake/structured_object_store.py [hunk 1]
```diff
@@ -1223,6 +1241,46 @@ def _encode_structured_field(value: Any) -> tuple[dict[str, Any], Any]:
...
+        original_view = memoryview(original)
+        encoded_view = memoryview(encoded)
+    except TypeError:
+        return False
+    if not original_view.contiguous or not encoded_view.c_contiguous:
+        return False
+    if original_view.readonly or encoded_view.readonly:
+        return False
+    if original_view.nbytes != encoded_view.nbytes:
+        return False
```
[the remark this answers is not in the document]
Comment (yokinoshitayoki) [reply mooncake-wheel/mooncake/structured_object_store.py]: @zxpdemonio do we need to consider these suggestions?

File: mooncake-wheel/mooncake/structured_object_store.py [hunk 2]
```diff
@@ -225,15 +228,20 @@ def put_structured_object(
         chunk_bytes: Optional[int],
         policy: Optional[BundleTransferPolicy],
         max_inflight_put: Optional[int],
+        pre_registered_buffers: Optional[Mapping[str, bool]],
     ) -> RemoteBundleRef:
         metadata, buffers = _encode_structured_fields(payload.metadata, payload.buffers)
+        _validate_pre_registered_structured_buffers(
```
Comment (yokinoshitayoki) [inline mooncake-wheel/mooncake/structured_object_store.py]: Could we resolve the effective transfer policy before this validation and skip the pre-registered-buffer check when `copy_mode == "copy"`?

In copy mode, `put_payload_chunks()` immediately routes to `_put_chunks_direct()` and never consumes `pre_registered_buffers`, so validating the pre-registered source here can reject a buffer that the requested copy path would otherwise handle correctly. For example, a caller falling back with `policy=BundleTransferPolicy(copy_mode="copy")` while still passing `pre_registered_buffers={"payload": True}` will fail for a non-C-contiguous ndarray, even though no registration or `batch_put_from` will be used. This makes the explicit copy fallback fragile for callers that reuse the same registration map across fast-path and fallback attempts.

**Review by yokinoshitayoki — APPROVED** [review]
Thanks, this PR looks good to me now.
